src/object/obj_utils.c

```diff
@@ -86,6 +90,157 @@ daos_iods_free(daos_iod_t *iods, int nr, bool need_free)
 		D_FREE(iods);
 }
 
+int
+obj_latency_tm_init(uint32_t opc, int tgt_id, struct d_tm_node_t **tm, char *op,
+		    char *desc, bool server)
+{
+	unsigned int	bucket_max = 256;
+	int		i;
+	int		rc = 0;
+
+	for (i = 0; i < NR_LATENCY_BUCKETS; i++) {
+		char *path;
+
+		if (server) {
+			if (bucket_max < 1024) /** B */
+				D_ASPRINTF(path, "io/latency/%s/%uB/tgt_%u",
+					   op, bucket_max, tgt_id);
+			else if (bucket_max < 1024 * 1024) /** KB */
+				D_ASPRINTF(path, "io/latency/%s/%uKB/tgt_%u",
+					   op, bucket_max / 1024, tgt_id);
+			else if (bucket_max <= 1024 * 1024 * 4) /** MB */
+				D_ASPRINTF(path, "io/latency/%s/%uMB/tgt_%u",
+					   op, bucket_max / (1024 * 1024), tgt_id);
+			else /** >4MB */
+				D_ASPRINTF(path, "io/latency/%s/GT4MB/tgt_%u",
+					   op, tgt_id);
+		} else {
+			pid_t pid = getpid();
+			unsigned long tid = pthread_self();
+
+			if (bucket_max < 1024) /** B */
+				D_ASPRINTF(path, "%s/%u/%lu/io/latency/%s/%uB",
+					   dc_jobid, pid, tid, op, bucket_max);
+			else if (bucket_max < 1024 * 1024) /** KB */
+				D_ASPRINTF(path, "%s/%u/%lu/io/latency/%s/%uKB",
+					   dc_jobid, pid, tid, op, bucket_max / 1024);
+			else if (bucket_max <= 1024 * 1024 * 4) /** MB */
+				D_ASPRINTF(path, "%s/%u/%lu/io/latency/%s/%uMB",
+					   dc_jobid, pid, tid, op, bucket_max / (1024 * 1024));
+			else /** >4MB */
+				D_ASPRINTF(path, "%s/%u/%lu/io/latency/%s/GT4MB",
+					   dc_jobid, pid, tid, op);
+
+		}
+		rc = d_tm_add_metric(&tm[i], D_TM_STATS_GAUGE, desc, "us", path);
+		if (rc)
+			D_WARN("Failed to create per-I/O size latency "
+			       "sensor: "DF_RC"\n", DP_RC(rc));
+		D_FREE(path);
+
+		bucket_max <<= 1;
+	}
+
+	return rc;
+}
+
+void
+obj_metrics_free(void *data)
+{
+	D_FREE(data);
+}
+
+int
+obj_metrics_count(void)
+{
+	return (sizeof(struct obj_pool_metrics) / sizeof(struct d_tm_node_t *));
+}
+
+
+void *
+obj_metrics_alloc_internal(const char *path, int tgt_id, bool server)
+{
+	struct obj_pool_metrics	*metrics;
+	uint32_t		opc;
+	int			rc;
+
+	D_ASSERT(tgt_id >= 0);
+
+	D_ALLOC_PTR(metrics);
+	if (metrics == NULL)
+		return NULL;
+
+	/** register different per-opcode counters */
+	for (opc = 0; opc < OBJ_PROTO_CLI_COUNT; opc++) {
+		/** Then the total number of requests, of type counter */
+		rc = d_tm_add_metric(&metrics->opm_total[opc], D_TM_COUNTER,
+				     "total number of processed object RPCs",
+				     "ops", "%s/ops/%s/%s%u", path,
```

Comment:
Unfortunately, this leads to some strange client metric names, e.g. `pool/$uuid/ops/ec_agg/12345`. I changed the logic a little to only append the tgt_id path for server metrics in this commit: https://github.com/daos-stack/daos/pull/13545/commits/abe1830117bc3b28c88bce10fb7b3fc77b03d03e

That commit also fixes a bug where the server metrics are created without the tgt_id.

Reply:
Thanks.